src/Common/src/System/Net/Security/CertificateHelper.cs

```diff
@@ -33,39 +34,52 @@ internal static X509Certificate2 GetEligibleClientCertificate(X509Certificate2Co
                 return null;
             }
 
-            // Build a new collection with certs that have a private key. We need to do this manually because there is
-            // no X509FindType to match this criteria.
-            // Find(...) returns a collection of clones instead of a filtered collection, so do this before calling
-            // Find(...) to minimize the number of unnecessary allocations and finalizations.
-            var eligibleCerts = new X509Certificate2Collection();
             foreach (X509Certificate2 cert in candidateCerts)
             {
-                if (cert.HasPrivateKey)
+                // Must have private key.
+                if (!cert.HasPrivateKey)
                 {
-                    eligibleCerts.Add(cert);
+                    continue;
                 }
-            }
+                
+                // If an extension is missing then it will automatically match.
+                bool isMatch = true;
 
-            // Don't call Find(...) if we don't need to.
-            if (eligibleCerts.Count == 0)
-            {
-                return null;
-            }
+                foreach (X509Extension extension in cert.Extensions)
+                {
+                    if (extension is X509EnhancedKeyUsageExtension ekus)
+                    {
+                        isMatch = false;
 
-            // Reduce the set of certificates to match the proper 'Client Authentication' criteria.
-            // Client EKU is probably more rare than the DigitalSignature KU. Filter by ClientAuthOid first to reduce
-            // the candidate space as quickly as possible.
-            eligibleCerts = eligibleCerts.Find(X509FindType.FindByApplicationPolicy, ClientAuthenticationOID, false);
-            eligibleCerts = eligibleCerts.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, false);
+                        foreach (Oid oid in ekus.EnhancedKeyUsages)
```

Comment:
FWIW, I think it'd be easier to follow all of this logic, what's breaking out of what, etc., if the inner loops were refactored into their own helper methods.

Reply:
Ok.  I will refactor and submit new commit.

Reply:
Yeah, making this block be something like

```C#
if (extension is X509EnhancedKeyUsageExtension eku)
{
    isMatch = HasEku(eku, ClientAuthenticationOID);
}
```

would probably help clarity.

Reply:
Thx.  I've refactored the code to make it clearer in my latest commit.